ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/TransactionService.scala

```diff
@@ -0,0 +1,707 @@
+          assert(
+            transactions.size == transactionsToRead,
+            s"$transactionsToRead should have been received but ${transactions.size} were instead")
+        }
+    }
+
+  private[this] val deduplicateCommands =
+    LedgerTest(
+      "TXDeduplicateCommands",
+      "Commands with identical command identifier and application identifier should be accepted and deduplicated") {
+      ledger =>
+        for {
+          party <- ledger.allocateParty()
+          request <- ledger.submitAndWaitRequest(party, Dummy(party).create.command)
+          _ <- ledger.submitAndWait(request)
+          _ <- ledger.submitAndWait(request)
+          transactions <- ledger.flatTransactions(party)
+        } yield {
+          assert(
+            transactions.length == 1,
+            s"Only one transaction was expected to be seen but ${transactions.length} appeared")
+        }
+    }
+
+  private[this] val rejectEmptyFilter =
+    LedgerTest(
+      "TXRejectEmptyFilter",
+      "A query with an empty transaction filter should be rejected with an INVALID_ARGUMENT status") {
+      ledger =>
+        for {
+          party <- ledger.allocateParty()
+          request = ledger.getTransactionsRequest(Seq(party))
+          requestWithEmptyFilter = request.update(_.filter.filtersByParty := Map.empty)
+          failure <- ledger.flatTransactions(requestWithEmptyFilter).failed
+        } yield {
+          assertGrpcError(failure, Status.Code.INVALID_ARGUMENT, "filtersByParty cannot be empty")
+        }
+    }
+
+  private[this] val completeOnLedgerEnd = LedgerTest(
+    "TXCompleteOnLedgerEnd",
+    "A stream should complete as soon as the ledger end is hit") { ledger =>
+    val transactionsToSubmit = 14
+    for {
+      party <- ledger.allocateParty()
+      transactionsFuture = ledger.flatTransactions(party)
+      _ <- Future.sequence(Vector.fill(transactionsToSubmit)(for {
+        contract <- ledger.create(party, Dummy(party))
+      } yield contract))
+      _ <- transactionsFuture
+    } yield {
+      // doing nothing: we are just checking that `transactionsFuture` completes successfully
+    }
+  }
+
+  private[this] val processInTwoChunks = LedgerTest(
+    "TXProcessInTwoChunks",
+    "Serve the complete sequence of transactions even if processing is stopped and resumed") {
+    ledger =>
+      val transactionsToSubmit = 5
+      for {
+        party <- ledger.allocateParty()
+        _ <- Future.sequence(Vector.fill(transactionsToSubmit)(for {
+          contract <- ledger.create(party, Dummy(party))
+        } yield contract))
+        endAfterFirstSection <- ledger.currentEnd()
+        firstSectionRequest = ledger
+          .getTransactionsRequest(Seq(party))
+          .update(_.end := endAfterFirstSection)
+        firstSection <- ledger.flatTransactions(firstSectionRequest)
+        _ <- Future.sequence(Vector.fill(transactionsToSubmit)(for {
```

Comment:
Same remark as above regarding the for-comprehension.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/2713/commits/c46eecc2eb97ad87d0f9212e3d76970c9376403f